python_modules/libraries/dagster-k8s/dagster_k8s/container_context.py

```diff
@@ -96,44 +112,80 @@ def __new__(
-            ).to_dict(),
+            server_k8s_config=server_k8s_config or UserDefinedDagsterK8sConfig.from_dict({}),
+            run_k8s_config=run_k8s_config or UserDefinedDagsterK8sConfig.from_dict({}),
             env=[
                 k8s_snake_case_dict(kubernetes.client.V1EnvVar, e)
                 for e in check.opt_sequence_param(env, "env")
             ],
         )
 
     def _merge_k8s_config(
-        self, first_config: Mapping[str, Any], second_config: Mapping[str, Any]
-    ) -> Mapping[str, Any]:
+        self,
+        onto_config: UserDefinedDagsterK8sConfig,
+        from_config: UserDefinedDagsterK8sConfig,
+    ) -> UserDefinedDagsterK8sConfig:
         # Keys are always the same and initialized in constructor
-        assert set(first_config) == set(second_config)
+        merge_behavior = from_config.merge_behavior
+        onto_dict = onto_config.to_dict()
+        from_dict = from_config.to_dict()
+        assert set(onto_dict) == set(from_dict)
+        if merge_behavior == K8sConfigMergeBehavior.DEEP:
+            onto_dict = copy.deepcopy(onto_dict)
+            merged_dict = self._deep_merge_k8s_config(onto_dict=onto_dict, from_dict=from_dict)
+        else:
+            merged_dict = {
+                key: (
+                    {**onto_dict[key], **from_dict[key]}
+                    if isinstance(onto_dict[key], dict)
+                    else from_dict[key]
+                )
+                for key in onto_dict
+            }
+        return UserDefinedDagsterK8sConfig.from_dict(merged_dict)
+
+    def _deep_merge_k8s_config(self, onto_dict: Dict[str, Any], from_dict: Mapping[str, Any]):
+        for from_key, from_value in from_dict.items():
+            if from_key not in onto_dict:
+                onto_dict[from_key] = from_value
+
+            else:
+                onto_value = onto_dict[from_key]
 
-        return {key: {**first_config[key], **second_config[key]} for key in first_config}
+                if (
+                    isinstance(from_value, list)
+                    and from_key not in ALWAYS_SHALLOW_MERGE_LIST_FIELDS
+                ):
+                    check.invariant(isinstance(onto_value, list))
+                    onto_dict[from_key] = _dedupe_list([*onto_value, *from_value])
+                elif isinstance(from_value, dict):
+                    check.invariant(isinstance(onto_value, dict))
+                    onto_dict[from_key] = self._deep_merge_k8s_config(onto_value, from_value)
+                else:
+                    onto_dict[from_key] = from_value
+        return onto_dict
 
-    def merge(self, other: "K8sContainerContext") -> "K8sContainerContext":
+    def merge(
+        self,
+        other: "K8sContainerContext",
+    ) -> "K8sContainerContext":
         # Lists of attributes that can be combined are combined, scalar values are replaced
         # prefering the passed in container context
```

Comment:
This comment is now a lie! (Mostly commenting to make it easy to follow this PR)